server/svix-server/src/core/security.rs

```diff
@@ -77,11 +128,83 @@ where
     }
 }
 
+pub struct AuthenticatedOrganization {
+    pub permissions: Permissions,
+}
+
+#[async_trait]
+impl<B> FromRequest<B> for AuthenticatedOrganization
+where
+    B: Send,
+{
+    type Rejection = Error;
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self> {
+        let permissions = Permissions::from_request(req).await?;
+        if permissions.type_ == KeyType::Application {
+            return Err(HttpError::unauthorized(
+                None,
+                Some(
+                    "You are only permitted to perform operations under the Application type"
+                        .to_owned(),
+                ),
+            )
+            .into());
+        }
+
+        Ok(AuthenticatedOrganization { permissions })
+    }
+}
+
 #[derive(Deserialize)]
 struct ApplicationPathParams {
     app_id: ApplicationIdOrUid,
 }
 
+pub struct OrganizationAuthenticatedApplication {
+    pub permissions: Permissions,
+    pub app: application::Model,
+}
+
+#[async_trait]
+impl<B> FromRequest<B> for OrganizationAuthenticatedApplication
+where
+    B: Send,
+{
+    type Rejection = Error;
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self> {
+        let permissions = Permissions::from_request(req).await?;
+
+        if permissions.type_ == KeyType::Application {
+            return Err(HttpError::unauthorized(
+                None,
+                Some(
+                    "You are only permitted to perform operations under the Application type"
+                        .to_owned(),
+                ),
+            )
+            .into());
+        }
+
+        let Path(ApplicationPathParams { app_id }) =
+            Path::<ApplicationPathParams>::from_request(req)
+                .await
+                .map_err(to_internal_server_error)?;
+        let Extension(ref db) = Extension::<DatabaseConnection>::from_request(req)
+            .await
+            .map_err(to_internal_server_error)?;
+        let app = application::Entity::secure_find_by_id_or_uid(
+            permissions.org_id.clone(),
+            app_id.to_owned(),
+        )
+        .one(db)
+        .await?
+        .ok_or_else(|| HttpError::not_found(None, None))?;
+        Ok(OrganizationAuthenticatedApplication { permissions, app })
+    }
+}
+
 pub struct AuthenticatedApplication {
     pub permissions: Permissions,
     pub app: application::Model,
```

Comment:
I can't make a comment on the correct line because Github sucks, though you need to add a check to verify that the app_id in the JWT is either None, or is the same as the path's app_id.